design-docs/improved-console.md

```diff
@@ -0,0 +1,422 @@
+ * `ProgressBarRegion` prefix
+ * `ProgressBarRegion` suffix
+ * `ProgressBar` complete char (default is "#")
+ * `ProgressBar` width (default is 50)
+
+### Implementation
+`ProgressBarFormatter` implementation would read the following properties and provide defaults.
+```bash
+org.gradle.console.progressbar.succeeded.prefix="[32m"   #FG_GREEN
+org.gradle.console.progressbar.failed.prefix="[31m"      #FG_RED
+org.gradle.console.progressbar.inprogress.prefix="[33m"  #FG_YELLOW
+org.gradle.console.progressbar.unstarted.prefix="[30m"   #FG_BLACK
+org.gradle.console.progressbar.region.suffix=""
+org.gradle.console.progressbar.completechar="#"
+org.gradle.console.progressbar.width=50
+```
+
+### Test Cases
+* Helpful error message is displayed if progressbar.width isn't a natural number
+* `progressbar.completechar` accepts multi-byte characters and displays correctly
+* Width of progress bar is consistent given multi-character "suffix"
+* Helpful error message if `progressbar.completechar` isn't one character in length
+* Progress bar is truncated at width of Console
+* Fuzz testing to protect against any security issues arising from using user input
+
+### Open issues
+* Could this concept allow customization of ProgressLine format, providing even greater flexibility?
+
+## Story: Gradle-provided slick themes
+Configure the default `org.gradle.console.progressbar.*` prefix and suffix based on a single 
+property.
+
+For example, a [powerline](https://github.com/powerline/powerline) theme, we would change the
+default progress bar properties to:
+
+```bash
+org.gradle.console.progressbar.succeeded.prefix="[42m"   #BG_GREEN
+org.gradle.console.progressbar.failed.prefix="[41m"      #BG_RED
+org.gradle.console.progressbar.inprogress.prefix="[43m"  #BG_YELLOW
+org.gradle.console.progressbar.unstarted.prefix="[40m"   #BG_BLACK
+org.gradle.console.progressbar.region.suffix=""
+org.gradle.console.progressbar.completechar=" "
+```
+
+### Implementation
+A `PowerlineProgressBarRegion` would extend `ProgressBarRegion` with the defaults
+given above. `BuildProgressBarListener` would detect the use of a Gradle property
+`org.gradle.console.theme=powerline` and substitute `PowerlineProgressBarRegion` 
+instead of the `DefaultProgressBarRegion`.
+
+## Story: (Optional) Customizable StatusBarFormat
+Allow users to configure the format of `StatusLine` entries in the status TextArea.
+
+Aspects that can be customized:
+ * `StatusBarLine` Prefix
+ * `StatusBarLine` Suffix
+
+### Implementation
+Introduce Gradle properties that allow this customization (with defaults):
+```bash
+org.gradle.console.statusline.prefix=" > "
+org.gradle.console.statusline.suffix=""
+```
+
+### Test Cases
+* Fuzz testing ensures no other parts of build are affected by user input
+
+### Open issues
+* How is intra-operation progress ([55 / 1234] or spinner) affected?
+
+## Story: (Optional) Indicate Gradle's continued progress in absence of intra-operation updates
```

Comment:
A couple notes about this. I think you're almost certainly right that we need this. The thing that sucks is that everything that looks polished isn't compatible with nearly all terminal environments (console size, color, font, font-size, etc). I'll think harder about how this might look, but so far I've only come up with ugly solutions.